bulkio: Add scheduled backup implementation

miretskiy opened the pull request:
Informs #28266

Add initial implementation for the "CREATE SCHEDULE FOR BACKUP" statement.

Implement schedule backup executor responsible for actual execution of the
create schedule for backup statement.  The schedule backup executor
evaluates tree.ScheduledBackup representation of the
`create schedule for backup` statements, and invokes backup plan hook
responsible for executing the actual backup.

Release Notes (enterprise): Implement "create schedule for backup"
which allows the creation of periodic backup schedules.

Commit:
jobs: Move scheduled job daemon start to sql server.

Refactor/move scheduled job daemon start to sql server.
This move is done for consistency reason (that where jobs registry
starts) as well as the follow on changes need to pull in more
dependencies from the sql server.

Release Notes: None

Commit:
jobs: Refactor scheduled jobs environment interface into separate package.

Refactor JobScheduler and JobSchedulerTestEnv into separate packages.
This change makes these 2 environments visible to the external packages
(which is necessary to implement scheduled backup, etc).

In addition, the JobSchedulerEnv is moved into scheduled_job package
in order to resolve circular dependencies when trying to use
this environment from the external packages that may in turn depend
on the jobs package.

Similiar, the JobSchedulerTestEnv is moved into jobstest package so that
it can be used in the external tests.

Release Notes: None.

Commit:
jobs: Introduce scheduled jobs JobExecutionConfig

Introduce scheduledjobs.JobExecutionConfig struct encapsulating the
state necessary for scheduled job execution (s.a. internal executor, etc).

Update scheduled job APIs to use this new struct.

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51308)
<!-- Reviewable:end -->

pbardea reviewed (commented):
Minor comments on my first pass - but planning on taking another pass through. Approach generally looks good to me though

> Reviewed 2 of 2 files at r1, 8 of 8 files at r2, 10 of 10 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-:-MC9P1P_Ao2iG1j8fsUT:bhzqjf7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 139 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9LELc04fz-Ctkdy2d:-MC9LELc04fz-Ctkdy2e:b-7l9zl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L139)):*
> ```Go
> 
> // doCreateBackupSchedule creates requested schedule (or schedules).
> // It is a plan hook implementation responsible for for the creating of scheduled backup.
> ```

nit: for for
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 343 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9LVuc9kPecAWCU9C9:-MC9LVuc9kPecAWCU9CA:b-ddgvn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L343)):*
> ```Go
> 	if !backupStmt.Options.Detached {
> 		backupStmt.Options.Detached = true
> 		log.Errorf(ctx, "force setting detached option for backup schedule %d", sj.ScheduleID())
> ```

This feels more like log.Warning rather than an error to me.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 376 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9MFIp-xTHXw_GKXk4:-MC9MFIp-xTHXw_GKXk5:bkfyq8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L376)):*
> ```Go
> 	}
> 	if planBackup == nil {
> 		return errors.Newf("backup evaluation error")
> ```

It may be useful to add the `backupStmt` here?
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9NWE3CbZEorOc-_WS:-MC9NWE3CbZEorOc-_WT:bd05h1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L382)):*
> ```Go
> 	}
> 
> 	resultCh := make(chan tree.Datums) // No need to close
> ```

Hmm, why is there no need to close it?
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 402 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9Nl_U6JRrti2tGOUB:-MC9Nl_U6JRrti2tGOUC:bc1uln7) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L402)):*
> ```Go
> 	g.GoCtx(func(ctx context.Context) error {
> 		if err := planBackup(ctx, nil, resultCh); err != nil {
> 			return errors.Wrap(err, "backup execution error")
> ```

Would "backup planning error" make it more clear here that the error did not happen during the job execution.
___
*[pkg/jobs/job_scheduler.go, line 306 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9OYcD2BsUVwQh0Wpx:-MC9OYcD2BsUVwQh0Wpy:b-2dzxp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/jobs/job_scheduler.go#L306)):*
> ```Go
> 	}
> 	if n != 1 {
> 		return errors.Newf("expected to update 1 row, update %d for job %d", n, jobID)
> ```

nit: `update_d_ %d`


<!-- Sent from Reviewable.io -->

Commit:
bulkio: Add scheduled backup implementation.

Add initial implementation for the "CREATE SCHEDULE FOR BACKUP" statement.

Implement schedule backup executor responsible for actual execution of the
create schedule for backup statement.  The schedule backup executor
evaluates tree.ScheduledBackup representation of the
`create schedule for backup` statements, and invokes backup plan hook
repsonsible for executing the actual backup.

Release Notes (enterprise): Implement "create schedule for backup"
which allows the creation of periodic backup schedules.

dt reviewed (commented):
> Reviewed 2 of 2 files at r1, 8 of 8 files at r2, 10 of 10 files at r3, 11 of 14 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/backup.proto, line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCAOZUN4unuvV17nEfK:-MCAOZUN4unuvV17nEfL:by6fkfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/backup.proto#L164)):*
> ```protobuf
>   BackupType backup_type = 1;
>   string backup_statement = 2;
> }
> ```

nit: newline
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9NWE3CbZEorOc-_WS:-MCAR-ESDZeesiX4Z0_D:b26qt57) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L382)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Hmm, why is there no need to close it?
</blockquote></details>

presumably since we stop waiting on it once we've read one row or ctx closes. as along as there's no way `planBackup` doesn't push one row to it without returning an err that would make the group close the ctx, it should be fine.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 412 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCAPyOVDNd9e-N2Grzh:-MCAPyOVDNd9e-N2Grzi:blxvj43) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L412)):*
> ```Go
> 
> 	// Mark the job as created by this schedule.
> 	if err := jobs.MarkJobCreatedBy(
> ```

hm. I wonder if we could / should just put this either in some sort of session or ctx, or even stuff it in the AST, so that planBackup could just create the job with this set correctly the first time? I guess we're doing it in same txn so i suppose it doesn't matter that much, but it seems weird to UPDATE a row we just created in the same txn? (also, we need to _really_ audit backup planning txn use either way, but especially if we're going to rely on this update).


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/base/testing_knobs.go, line 35 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCZBUb4q4O6RVEGoAe:-MCCZBUb4q4O6RVEGoAf:b3qv5nh) ([raw file](https://github.com/cockroachdb/cockroach/blob/52334bf2e9ec0e771c3c5a0dd97cc1b4ed8b8c35/pkg/base/testing_knobs.go#L35)):*
> ```Go
> 	Server              ModuleTestingKnobs
> 	TenantTestingKnobs  ModuleTestingKnobs
> 	JobsTestingKnobs    ModuleTestingKnobs
> ```

FYI: I had to add JobsTestingKnobs both to base.TestingKnobs as well as execinfra.TestingKnobs

The reason is that plan hooks have access to execinfra.TestingKnobs, but not base.TestingKnobs.
On the other hand, I needed to use base.TestingKnobs when testing job scheduler.  However, jobs cannot depend on execinfra.  So, the tests initialize base.TestingKnobs knobs; and I ensure that execinfra.TestingKnobs.JobsTestingKnobs == base.TestingKnobs.JobsTestingKnobs.
___
*[pkg/ccl/backupccl/backup.proto, line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCAOZUN4unuvV17nEfK:-MCCUx9JAEky2sN0Y8_0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/backup.proto#L164)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: newline
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 139 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9LELc04fz-Ctkdy2d:-MCCV5wUE3ARc61Op_DP:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L139)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

nit: for for
</blockquote></details>

Done
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 343 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9LVuc9kPecAWCU9C9:-MCCV7k9--KJrALpwwqc:b-r3q1id) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L343)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

This feels more like log.Warning rather than an error to me.
</blockquote></details>

Okay... It's more of a precondition really (since I set detached when I serialize).
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 376 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9MFIp-xTHXw_GKXk4:-MCCVqhY9lKyRgT0_x4x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L376)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

It may be useful to add the `backupStmt` here?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9NWE3CbZEorOc-_WS:-MCCVt4M6uDKTnyGc6nq:b-7uugl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L382)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

presumably since we stop waiting on it once we've read one row or ctx closes. as along as there's no way `planBackup` doesn't push one row to it without returning an err that would make the group close the ctx, it should be fine.
</blockquote></details>

Right... Basically, close on a channel is useful when you want to tell consumers that there is no more data coming.  Since I run both producer and consumer, I know exactly when there is no more data (i.e. after planBackup() returns).
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 402 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9Nl_U6JRrti2tGOUB:-MCCWFSq6xZgX3Vbo9Fu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L402)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Would "backup planning error" make it more clear here that the error did not happen during the job execution.
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 412 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCAPyOVDNd9e-N2Grzh:-MCCYEusDGwMGf2MU89v:b-4voxvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L412)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

hm. I wonder if we could / should just put this either in some sort of session or ctx, or even stuff it in the AST, so that planBackup could just create the job with this set correctly the first time? I guess we're doing it in same txn so i suppose it doesn't matter that much, but it seems weird to UPDATE a row we just created in the same txn? (also, we need to _really_ audit backup planning txn use either way, but especially if we're going to rely on this update).
</blockquote></details>

I agree.  I would have wanted to have jobs system set these appropriately, but I did not want to get into changing how we create jobs as part of this PR...
It is weird, but as you said, because under txn -- it's okay.  I added todo on MarkJobCreatedBy to remove it/rework it.
___
*[pkg/jobs/job_scheduler.go, line 227 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCC_WvK25ucSR2jO081:-MCC_WvLDAXyoA1Yo7Ed:blazc3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/52334bf2e9ec0e771c3c5a0dd97cc1b4ed8b8c35/pkg/jobs/job_scheduler.go#L227)):*
> ```Go
> 
> // Returns the amount of time to wait before starting initial scan.
> func getInitialScanDelay(knobs base.ModuleTestingKnobs) time.Duration {
> ```

FYI: I had to change getInitialScanDelay and getWaitPeriod to use testing knobs.
The reason is that modifying these variables was racy (as discovered by -race flag) when I was modifying those settings in a the test, while the real daemon was still running.

This necessitated reworking of some of the testHelper logic.


<!-- Sent from Reviewable.io -->

pbardea reviewed (approved):
> Reviewed 3 of 14 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCXlFEBv5lOQ8iwAb0:-MCCXlFEBv5lOQ8iwAb1:b-t88fpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L225)):*
> ```Go
> 	// TODO(yevgeniy): Validate backup schedule:
> 	//  * Verify targets exist.  Provide a way for user to override this via option.
> 	//  * Verify destination paths sane (i.e. valid schema://, etc)
> ```

Not for this PR, but I wonder if it would make sense to have some sort of planning "dry-run" that can do this validation. It would make schedule creation a bit more expensive, but it would also give the user some useful information (e.g. do I have access to the destination bucket?)
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9NWE3CbZEorOc-_WS:-MCCaMZ35dUBtJzv1KrC:b-dc62dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L382)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Right... Basically, close on a channel is useful when you want to tell consumers that there is no more data coming.  Since I run both producer and consumer, I know exactly when there is no more data (i.e. after planBackup() returns).
</blockquote></details>

:+1: thanks
___
*[pkg/ccl/backupccl/create_scheduled_backup_test.go, line 317 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCZ56901S8fYGEFSIm:-MCCZ56901S8fYGEFSIn:b-851u5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup_test.go#L317)):*
> ```Go
> 				}))
> 
> 			// Wait for the backup to complete
> ```

nit: `.` at end of line
___
*[pkg/jobs/testing_knobs.go, line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCaSSQ0vK4zXxt2jiG:-MCCaSSR6NWXHEUuMe6V:b-29siez) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/jobs/testing_knobs.go#L29)):*
> ```Go
> 	// the normal job scheduler daemon logic.
> 	// This function will be passed in a function which the caller
> 	// may invoke directly, bypassing normal job scheduler daemon logic.
> ```

nit: comment wrapping (on the first sentence)
___
*[pkg/sql/execinfra/server_config.go, line 231 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCYxG21Um4t-kmp-0h:-MCCYxG21Um4t-kmp-0i:b-851u5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/sql/execinfra/server_config.go#L231)):*
> ```Go
> 	BulkAdderFlushesEveryBatch bool
> 
> 	// JobsTestingKnobs is jobs infra specific testing knobs
> ```

nit: `.` at end of line


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCXlFEBv5lOQ8iwAb0:-MCCdhmv6U9t4sqkLzMk:b-128vje) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L225)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Not for this PR, but I wonder if it would make sense to have some sort of planning "dry-run" that can do this validation. It would make schedule creation a bit more expensive, but it would also give the user some useful information (e.g. do I have access to the destination bucket?)
</blockquote></details>

I'm not sure we would need separate dry run.  We have it already:
BEGIN
CREATE SCHEDULE FOR .....
ROLLBACK
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 343 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MC9LVuc9kPecAWCU9C9:-MCCdeVw4JhJ4LE0Jlaa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L343)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Okay... It's more of a precondition really (since I set detached when I serialize).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCXlFEBv5lOQ8iwAb0:-MCCi0RbE4oZuXpeaa81:b-ke61t0) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L225)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not sure we would need separate dry run.  We have it already:
BEGIN
CREATE SCHEDULE FOR .....
ROLLBACK
</blockquote></details>

Sorry, I didn't mean having a "dry-run" from the user's perspective, I meant that when creating the schedule we could execute the backup plan hook now, rather than waiting for the first scheduled run but don't create the job. That way all of the validation would be done in backup planning. Things may get a bit tricky though since we're touching external resources inside a transaction (but we should only be reading them).


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MCCXlFEBv5lOQ8iwAb0:-MCClVRv1VB4aOUYEscv:b-gzlhgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/de99f7a2c243e839af3907c8600e760bbd2a119f/pkg/ccl/backupccl/create_scheduled_backup.go#L225)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Sorry, I didn't mean having a "dry-run" from the user's perspective, I meant that when creating the schedule we could execute the backup plan hook now, rather than waiting for the first scheduled run but don't create the job. That way all of the validation would be done in backup planning. Things may get a bit tricky though since we're touching external resources inside a transaction (but we should only be reading them).
</blockquote></details>

I'm still not too sure if it makes sense.  We could validate e.g. that TO has valid schemas on the uris, etc... but beyond that, any validation we may do now, might not be very useful since conditions (in the external world) may change before you run the schedule anyway.


<!-- Sent from Reviewable.io -->

dt commented on pkg/ccl/backupccl/create_scheduled_backup_test.go:
```diff
@@ -0,0 +1,328 @@
+			// Verify backup.
+			th.sqlDB.Exec(t, "CREATE DATABASE restore_db; USE restore_db")
+			defer th.sqlDB.Exec(t, "DROP DATABASE restore_db")
+			th.sqlDB.Exec(t, tc.restore, destination)
```
We could just run `SELECT table_name FROM [SHOW BACKUP $1]` and verify we get back the set of verify tables -- we only need to be testing that the schedule backed up what it was supposed to backup, not that backup and restore work since that's the responsibility of the backup tests.

dt commented on pkg/ccl/backupccl/create_scheduled_backup.go:
```diff
@@ -0,0 +1,466 @@
+	if err != nil {
+		return nil, nil, nil, false, err
+	}
+	return func(ctx context.Context, _ []sql.PlanNode, resultsCh chan<- tree.Datums) error {
```
minor readability nit: i'd do `fn := ...`  \n `return fn ...`

dt commented on pkg/ccl/backupccl/create_scheduled_backup.go:
```diff
@@ -0,0 +1,466 @@
+	return nil, errors.Newf("unexpect node type %T", node)
+}
+
+type scheduledBackupExecutor struct{}
```
2¢: I think this could move to its own file `schedule_exec.go` or something.

dt commented on pkg/ccl/backupccl/create_scheduled_backup.go:
```diff
@@ -0,0 +1,466 @@
+	return nil
+}
+
+const scheduleBackupOp = "SCHEDULED_BACKUP"
```
Shouldn't this be `CREATE SCHEDULE FOR BACKUP`? Otherwise the errors below will read "argument of SCHEDULE_BACKUP must be .." which isn't a thing

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 262 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4uotFBlmfk7EUg-r12-262:-MCSCjp61a8PjksEJkky:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/ccl/backupccl/create_scheduled_backup.go#L262)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Shouldn't this be `CREATE SCHEDULE FOR BACKUP`? Otherwise the errors below will read "argument of SCHEDULE_BACKUP must be .." which isn't a thing
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 454 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4uotFBlmfk7EUg-r12-454:-MCSD0Uz7HRdNXUo3WBC:bjuenjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/ccl/backupccl/create_scheduled_backup.go#L454)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

minor readability nit: i'd do `fn := ...`  \n `return fn ...`
</blockquote></details>

Ty. Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup_test.go, line 82 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4uotFBlmfk7EUh-r12-82:-MCSDCWr0PCTwCUerMNK:b-i0hy6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/ccl/backupccl/create_scheduled_backup_test.go#L82)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Eh, I'd avoid session vars since they play poorly with connection pools in weird and unexpected ways. I'd just put `WHERE true` in your `DELETE`s to shut it up.
</blockquote></details>

Neat.  Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup_test.go, line 318 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4uotFBlmfk7EUh-r12-318:-MCSf7JF0yRHFwOQhS3-:bexk120) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/ccl/backupccl/create_scheduled_backup_test.go#L318)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Do we need the session DB set?
</blockquote></details>

no longer applicab le.
___
*[pkg/ccl/backupccl/create_scheduled_backup_test.go, line 320 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4uotFBlmfk7EUh-r12-320:-MCS_Rj09okfNmG6fR9n:b2ody5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/ccl/backupccl/create_scheduled_backup_test.go#L320)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We could just run `SELECT table_name FROM [SHOW BACKUP $1]` and verify we get back the set of verify tables -- we only need to be testing that the schedule backed up what it was supposed to backup, not that backup and restore work since that's the responsibility of the backup tests.
</blockquote></details>

Good point.  Done.
___
*[pkg/jobs/job_scheduler.go, line 189 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4sLxpxUDgdnVCH-r12-189:-MCSf5PbCBVThCPMFlM_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/jobs/job_scheduler.go#L189)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think we usually prefer to make a timer once and reset it it each loop instead of calling time.After every iteration and having it allocate a new channel each time.  

This usually looks like 
```
var timer timeutil.Timer
defer timer.Stop()
for {
  howLongToWait := ...
  timer.Reset(howLongToWait)
  select {
   cases <-ctxDone/ShouldStop/etc
   case <-timer.C:
      timer.Read = true
  }
 doTheThing() 
}
```
</blockquote></details>

Done.
___
*[pkg/server/server_sql.go, line 670 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4sLxpxUDgdnVCG-r12-670:-MCSfN4aC1K4kdcI6UAS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/server/server_sql.go#L670)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This indentation is a bit weird. I think we usually do "whole-call on one line" -> "all the args on one line, with call opening above, closing below" -> "each arg starting on its own line". I think that last one would be a bit more readable here with this big multi-line arg.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51308)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 325 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51308#-MBuJ-4uotFBlmfk7EUg-r12-325:-MCSgtl85HNmPIGlJg3u:b-eiuyuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a91db0378b91d3f3192162ed76700c1bb0cc191/pkg/ccl/backupccl/create_scheduled_backup.go#L325)):*
> schedule_exec.go 


<!-- Sent from Reviewable.io -->

miretskiy commented:
bors r+
